doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1457,13 +1479,14 @@ expression returns an identifier:
 
 .. tacn:: fresh {* {| @string | @qualid } }
 
-   Evaluates to an identifier unbound in the goal. This fresh identifier
-   is formed by concatenating the value of each :n:`@qualid`,
-   which has to refer to an unqualified name, and :n:`string`.
+   .. todo you can't have a :tacn: with the same name as a :gdef: for now,
+      eg `fresh` can't be both
 
-   If the resulting name is already used, it is padded with a number so that it
-   becomes fresh. If no component is given, the name is a fresh derivative of
-   the name ``H``.
+   Returns a fresh identifier name (i.e. one that is not already used in the context).
```

Comment:
Not sure if it's worth talking about here, but this context is the goal context concatenated with the tactic-local Ltac context.  For example
```coq
Goal True -> True.
  intro x.
  let f := fun _ => fresh "x" in
  let a := fresh "x" in
  let b := fresh "x" in
  let c := fresh "x" in
  let f2 := fun _ => fresh "x" in
  let d := f () in
  let e := f () in
  let g := f2 () in
  let h := f2 () in
  idtac a b c d e g h. (* x0 x1 x2 x0 x0 x3 x3 *)
```
This is very subtle behavior that might trip up authors of Ltac automation (and has in fact tripped me up).

Reply:
I think it's worth including.  Seems a bit of a bug, not a useful feature.

I take it the "fun" saves the state of the context, so it a "fresh" within will always give the same value?  But a global definition using "fresh" will work as expected?  

And there is some tracking of what fresh has returned so a, b and c are different even though x0 x1 x2 are not in the context yet?

What can we recommend as best practice to avoid this?

Reply:
@JasonGross Any response?

Reply:
> I think it's worth including. Seems a bit of a bug, not a useful feature.

Yep
 
> I take it the "fun" saves the state of the context, so it a "fresh" within will always give the same value?

Yes.

> But a global definition using "fresh" will work as expected?

No, the same issue shows up in global Ltac definitions which are recursive; invocations of `fresh` in the recursive call aren't aware of the invocation of `fresh` in the parent caller.

> And there is some tracking of what fresh has returned so a, b and c are different even though x0 x1 x2 are not in the context yet?

Yes

> What can we recommend as best practice to avoid this?

Always use Ltac2?  (Ltac2 lets you be explicit about which identifiers you're avoiding.)  Make the `fresh` and the corresponding `intro`/`pose`/whatever is using the name be close together, with no intervening calls to anything that might generate names?

Reply:
First: to apply a function, you must provide at least one argument, right?

Is this correct?

> A function application is an expression of the form:

> .. tacn:: @qualid {+ @tactic_arg }

>   :n:`@qualid` must be bound to a |Ltac| function
   with at least as many arguments as the provided :n:`@tactic_arg`\s.
   The :n:`@tactic_arg`\s are evaluated before the function is applied
   or partially applied.
   
>   Note that functions may be defined with both :tacn:`fun` and :cmd:`Ltac`.

Second: it seems global assignments such as `Ltac a := 1` are not useful,  right?  If you define 
a global function, that work, not sure if that's often useful.  And `Ltac x := intro` is fine.

```
  Ltac n := 1.
  idtac n.
(* n not found *)

  Ltac n := fun _ => 1.
  let n2 := n () in idtac n2.
(* 1 *)

  Ltac x := intro.
  x.
(* applies "intro" *)
```

Reply:
> First: to apply a function, you must provide at least one argument, right?

This is a question of how you define function.  As I see it, there's only a presentational question here, not a technical one, and I'd like to defer to you and @Zimmi48 about this.  (If there is in fact a technical question, let me know, e.g., by giving two snippets of Ltac to pseudo-Ltac, and I can answer with my understanding.)

> :n:`@qualid` must be bound to a |Ltac| function
> with at least as many arguments as the provided :n:`@tactic_arg`\s.

This is almost correct, and the way it's incorrect might not be worth noting.  Consider the following example:
```coq
Ltac gobble0 x :=
  lazymatch x with
  | 0 => gobble0
  | _ => x
  end.

Goal True.
  let y := gobble0 0 0 1 in
  idtac y.
```
This is very silly, but how many arguments does `gobble0` have?  I'd say one, because the expression underneath the first `fun x => ...` is not another `fun`.  But it also sort-of has infinitely many arguments.  I do sometimes do things like
```coq
Ltac foo x :=
  let y := some_expensive_computation in
  fun z => ...
```
And then I'll bind `foo x` (or `foo x z`) to some identifier and then later apply it multiple times.



> Second: it seems global assignments such as `Ltac a := 1` are not useful, right?

By contrast, I use this to define configurable options for my scripts.  Consider:
```coq
Ltac verbose := constr:(false).

Ltac if_verbose tac :=
  lazymatch verbose with
  | true => tac ()
  | _ => idtac
  end.

Ltac do_stuff :=
  if_verbose ltac:(fun _ => idtac "intro");
  intro;
  if_verbose ltac:(fun _ => idtac "constructor");
  constructor.

Goal True -> True. do_stuff. Qed.
Ltac verbose ::= constr:(true).
Goal True -> True. do_stuff. Qed.
```
Very useful for, e.g., debugging reficiation tactics.

>     Ltac n := 1.
>     idtac n.
>     (* n not found *)

The correct version of this is
```coq
Ltac n := 1.
Goal True.  let nv := n in idtac nv. (* 1 *)
```

Reply:
>> First: to apply a function, you must provide at least one argument, right?
> This is a question of how you define function.

And what are those ways? I don't know what you have in mind.  If one argument is required, then I can resolve the apparent ambiguity described here https://github.com/coq/coq/pull/12103#issuecomment-639993027 by indicating at least one parameter is needed.

I'll pass on describing gobble for now.  It seems very unlike other languages, not easy to explain.

> The correct version of this is
> ```coq
> Ltac n := 1.
> Goal True.  let nv := n in idtac nv. (* 1 *)
```
So substitution only applies to let-defined and pattern-defined variables, not top-level symbols.

I was not expecting to start a long discussion here :-(.

Reply:
> And what are those ways? I don't know what you have in mind. If one argument is required, then I can resolve the apparent ambiguity described here [#12103 (comment)](https://github.com/coq/coq/pull/12103#issuecomment-639993027) by indicating at least one parameter is needed.

I think my point it that regardless of which way you resolve the parsing ambiguity, the semantics are the same, at least insofar as I'm aware.  So you can be like "yes, a function needs at least one argument, and then we have a separate case for top-level symbols" or you can be like "0-argument-functions are a thing, they behave just like top-level symbols", and it doesn't matter as far as understanding how Ltac programs are interpreted and run.

> I'll pass on describing gobble for now. It seems very unlike other languages, not easy to explain.

Yep, sounds good.  Adam pointed out to me in conversation today that you can actually get a similar effect in OCaml using ADTs (and of course you can get something like this in Gallina with dependent types), but I agree that it's not a central case in Ltac programming, and not worth describing at this point

> So substitution only applies to let-defined and pattern-defined variables, not top-level symbols.

This is true of substitution in `idtac`.  Probably also substitution in `fresh`.  This is also the case when you write `constr:(fun x => ...)`, where if `x` is an Ltac variable that has been let-bound, it'll look inside `x` for a name, but if `x` is a global Ltac, it'll just use the identifier `x`.  I have a feel for this corner of Ltac, but not a complete description of it, alas.  (I guess in some built-in constructs, "substitution of things in the local Ltac context" is used rather than more general Ltac expression-evaluation, and that's what's going on here.)

> I was not expecting to start a long discussion here :-(.

Yeah, Ltac is unpleasantly complicated.  I'd like to see this PR merged soon, too.